src/Functions/FunctionsConversion.h

```diff
@@ -1321,6 +1336,29 @@ struct ConvertThroughParsing
 
         size_t current_offset = 0;
 
+        bool precise_float_parsing = false;
+
+        if (DB::CurrentThread::isInitialized())
+        {
+            const DB::ContextPtr query_context = DB::CurrentThread::get().getQueryContext();
+
+            if (query_context)
+                precise_float_parsing = query_context->getSettingsRef().precise_float_parsing;
+        }
+
+        std::function<void(typename ToDataType::FieldType &, DB::ReadBuffer &, const DateLUTImpl *)> parseFunction;
+        std::function<bool(typename ToDataType::FieldType &, DB::ReadBuffer &, const DateLUTImpl *)> tryParseFunction;
+
+        if constexpr (!(parsing_mode == ConvertFromStringParsingMode::BestEffort ||
+              parsing_mode == ConvertFromStringParsingMode::BestEffortUS ||
+              to_datetime64 || IsDataTypeDecimal<ToDataType>))
+        {
+            if constexpr (exception_mode == ConvertFromStringExceptionMode::Throw)
+                parseFunction = precise_float_parsing ? parseImpl<true, ToDataType> : parseImpl<false, ToDataType>;
+            else
+                tryParseFunction = precise_float_parsing ? tryParseImpl<true, ToDataType> : tryParseImpl<false, ToDataType>;
+        }
```

Comment:
why we need `std::function`? this adds unnecessary overhead - can't we use `if constexpr` below?

Reply:
What do you mean?
This function is to be called later in that long if-constexpr (and it can only be initialized for float types), that's why they are declared separately.

Reply:
I mean instead creating these functions here and then calling them later - just call them directly later at 1453,1523 with `if constexpr`.
Also it makes impression that they can be not initialized because of the the first `if constexpr` at 1377...

Reply:
Precise_float_parsing isn't constexpr and otherwise it will be evaluated for each parsed value. This may have a small impact on parsing speed.

Reply:
Check how it is done for `date_time_input_format`.

Reply:
I see, it is evaluated in run time for every parsed value, e.g. [here](https://github.com/ClickHouse/ClickHouse/blob/ed40a15e82173b406ac6dfb895ac9ddf326941b1/src/DataTypes/Serializations/SerializationDateTime.cpp#L22).
If this is fine, I would get rid of this additional template parameter completely and the code would be clearer.
Maybe parsing itself is a very heavy operation and nothing bad in having one more `if` 😃

Reply:
> Also it makes impression that they can be not initialized because of the first `if constexpr` at 1377

Yes, they can. The same thing happens to `local_time_zone` and `utc_time_zone` at the beginning of the same `execute` function.

Reply:
yes, I missed that here `precise_float_parsing` isn't a template parameter... but either way maybe it's worth to just move `?:` to there - we call it for every row and std::function indirection can be more costly than simple condition...